libs/vkd3d/state.c

```diff
@@ -1189,36 +1194,36 @@ static HRESULT vkd3d_render_pass_cache_create_pass_locked(struct vkd3d_render_pa
         attachments[attachment_index].flags = 0;
         attachments[attachment_index].format = key->vk_formats[index];
         attachments[attachment_index].samples = key->sample_count;
-
-        if (key->depth_enable)
-        {
-            attachments[attachment_index].loadOp = VK_ATTACHMENT_LOAD_OP_LOAD;
-            attachments[attachment_index].storeOp = VK_ATTACHMENT_STORE_OP_STORE;
-        }
-        else
-        {
-            attachments[attachment_index].loadOp = VK_ATTACHMENT_LOAD_OP_DONT_CARE;
-            attachments[attachment_index].storeOp = VK_ATTACHMENT_STORE_OP_DONT_CARE;
-        }
-        if (key->stencil_enable)
-        {
-            attachments[attachment_index].stencilLoadOp = VK_ATTACHMENT_LOAD_OP_LOAD;
-            attachments[attachment_index].stencilStoreOp = VK_ATTACHMENT_STORE_OP_STORE;
-        }
-        else
-        {
-            attachments[attachment_index].stencilLoadOp = VK_ATTACHMENT_LOAD_OP_DONT_CARE;
-            attachments[attachment_index].stencilStoreOp = VK_ATTACHMENT_STORE_OP_DONT_CARE;
-        }
+        attachments[attachment_index].loadOp = VK_ATTACHMENT_LOAD_OP_LOAD;
+        attachments[attachment_index].storeOp = VK_ATTACHMENT_STORE_OP_STORE;
+        attachments[attachment_index].stencilLoadOp = VK_ATTACHMENT_LOAD_OP_LOAD;
+        attachments[attachment_index].stencilStoreOp = VK_ATTACHMENT_STORE_OP_STORE;
         attachments[attachment_index].initialLayout = depth_layout;
         attachments[attachment_index].finalLayout = depth_layout;
 
         attachment_references[index].attachment = attachment_index;
         attachment_references[index].layout = depth_layout;
 
+        stages |= VK_PIPELINE_STAGE_EARLY_FRAGMENT_TESTS_BIT | VK_PIPELINE_STAGE_LATE_FRAGMENT_TESTS_BIT;
         attachment_index++;
     }
 
+    dependencies[0].srcSubpass = VK_SUBPASS_EXTERNAL;
+    dependencies[0].dstSubpass = 0;
+    dependencies[0].srcStageMask = stages;
+    dependencies[0].dstStageMask = stages;
+    dependencies[0].srcAccessMask = 0;
+    dependencies[0].dstAccessMask = 0;
+    dependencies[0].dependencyFlags = 0;
```

Comment:
BY_REGION as well probably?

Reply:
Is `BY_REGION` meaningful at all outside of subpass self-dependencies?

Can do, at least AMD ignores this anyway, just not sure what exactly the semantics are supposed to be.

Reply:
Yes, it should allow a correct IMR implementation to elide the barrier entirely. Can't hurt to add it, and gives a reason for drivers to actually take BY_REGION into consideration.

Reply:
src/dstAccessMask being 0 here is technically wrong, but it's the only known way currently to avoid full stalls on drivers. BY_REGION is supposed to be the fix for it, but we can't rely on that yet.

Reply:
Alright, set `VK_DEPENDENCY_BY_REGION_BIT` and added a comment about the access masks.